executor/hnswlib_searcher.py

```diff
@@ -0,0 +1,360 @@
+        if embeddings.shape[-1] != self.dim:
+            raise ValueError(
+                f'Attempted to update vectors with dimension'
+                f' {embeddings.shape[-1]}, but dimension of index is {self.dim}'
+            )
+
+        self._index.add_items(embeddings, ids=doc_inds)
+
+    def delete(self, parameters: Dict, **kwargs):
+        """Delete entries from the index by id
+
+        :param parameters: parameters to the request. Should contain the list of ids
+            of entries (Documents) to delete under the `ids` key
+        """
+        deleted_ids = parameters.get('ids', [])
+
+        for _id in set(deleted_ids).intersection(self._ids_to_inds.keys()):
+            ind = self._ids_to_inds[_id]
+            self._index.mark_deleted(ind)
+            del self._ids_to_inds[_id]
+
+    def dump(self, parameters: Dict = {}, **kwargs):
+        """Save the index and document ids.
+
+        The index and ids will be saved separately for each shard.
+
+        :param parameters: Dictionary with optional parameters that can be used to
+            override the parameters set at initialization. The only supported key is
+            `dump_path`.
+        """
+
+        dump_path = parameters.get('dump_path', self.dump_path)
+        if dump_path is None:
+            raise ValueError(
+                'The `dump_path` must be provided to save the indexer state.'
+            )
+
+        self._index.save_index(f'{dump_path}/index.bin')
+        with open(f'{dump_path}/ids.json', 'w') as f:
+            json.dump(dict(self._ids_to_inds), f)
+
+    def clear(self, **kwargs):
+        """Clear the index of all entries."""
+        self._index = hnswlib.Index(space=self.metric_type, dim=self.dim)
+        self._init_empty_index()
+        self._index.set_ef(self.ef_query)
+
+    def status(self) -> Dict:
+        """Return the status information about the indexer.
+
+        The status will contain information on the total number of indexed and deleted
+        documents, and on the number of (searchable) documents currently in the index.
+        """
+
+        status = {
+            'count_deleted': self._index.element_count - len(self._ids_to_inds),
+            'count_indexed': self._index.element_count,
+            'count_active': self.size,
+        }
+        return status
+
+    @property
+    def size(self):
+        return len(self._ids_to_inds)
+
+    @property
+    def metric_type(self):
+        if self.metric == 'euclidean':
+            metric_type = 'l2'
+        elif self.metric == 'cosine':
+            metric_type = 'cosine'
+        elif self.metric == 'inner_product':
+            metric_type = 'ip'
+
+        if self.metric not in ['euclidean', 'cosine', 'inner_product']:
+            self.logger.warning(
+                f'Invalid distance metric {self.metric} for HNSW index construction! '
+                'Default to euclidean distance'
+            )
+            metric_type = 'l2'
```

Comment:
should use `cosine` as default, am I correct?

Reply:
Yes, done